### ISSUE TITLE
Finding the best preprocessing

### ISSUE THREAD

**Issue by Luugaaa** [ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star#24]
## Context

In order to improve the results, we want to find the best preprocessing for the magnitude and the phase. A good preprocessing should facilitate the segmentation for the model from the very beginning. To evaluate the best preprocessing, we conduct experiments based on the default implementation over 25 epochs. 

## Preprocessing techniques 

In the default implementation, there is no heavy preprocessing meant to improve the quality of the data given to the model. Here, for now, we investigate the impact of both the use CLAHE on the magnitude channel and/or the use of a linear contrast enhancement on the phase. We evaluate the impact on the validation metrics. Naturally, the preprocessing is applied on both train and val data and the training is deterministic. 

_Note for later : for now we implement the preprocessing as an additional step done after augmentation. For a better consistency, it could/should be integrated in the `get_training_transforms` and the `get_val_transforms`._

**Comment by Luugaaa** [comment]
Note : for the following experiments, the mag preprocessing is 
```python 
CLAHE(mag, clip_limit=0.01)
```
 and the phase preprocessing is : 
```python
p2, p98 = np.percentile(unwrapped_phase, (30, 75))
rescale_intensity(unwrapped_phase, in_range=(p2, p98), out_range=(0, 1))
```

First, we applied both the phase preprocessing and the magnitude preprocessing (red curve, Exp_2_default_prepro_mag_phase) and compared the results with the same implementation with no preprocessing applied (green curve, Exp_2_default_prepro_mag_phase). 

<img width="568" height="306" alt="Image" src="https://github.com/user-attachments/assets/e32b7ad6-6500-4beb-9e39-663491d7b174" />

At first the training with both prepro has worse results than that default training, but the gap seems to be closed toward the end. This is promising, we might want to test over more epochs. 

On each class we have the following results : 
* class 0 : bg
* class 1 : wm
* class 2 : gm
* class 3 : lesion_wm
* class 4 : lesion_gm

<img width="1148" height="628" alt="Image" src="https://github.com/user-attachments/assets/3c14c84a-8e20-413e-a22e-1d7b1a51ad17" />

Most difference is found in the segmentation of GM and WM. For GM, the preprocessing seems to greatly increase the results (+0.1 on dice), for the WM it's the other way around, we lose around 0.1 with the preprocessing. 

When removing the phase prepro, leaving only the magnitude prepro, there is a significant decrease in the mean dice (see below). 

<img width="569" height="305" alt="Image" src="https://github.com/user-attachments/assets/e4a0fcbf-f11c-4aae-af20-ae42070b9e59" />

For each class, both the dual preprocessing and the mag only preprocessing seems to yield similar results (see below), except for the GM segmentation. This seems to indicate the phase preprocessing is behind the improvement for the GM segmentation (which actually not really surprising). On the other hand, the WM segmentation stays low, this could indicate the mag preprocessing is the source of the dice decrease for this class.

<img width="1148" height="625" alt="Image" src="https://github.com/user-attachments/assets/2ee11192-4b57-41d4-98b5-1674a8c750b2" />

When preprocessing only the phase (orange curve, Exp_4_default_prepro_phase), the WM segmentation dice goes back to what is was without prepro, actually with a slight increase. Surprisingly, for the GM segmentation we don't observe the dice increase we had with both preprocessing. 

<img width="1151" height="626" alt="Image" src="https://github.com/user-attachments/assets/298031f6-39a2-4479-aee5-d534354828fb" />

The mean val dice is satisfying without showing any significant increase. 

<img width="568" height="306" alt="Image" src="https://github.com/user-attachments/assets/d317b25c-05bf-4cd6-9e3c-da35e56ee6f0" />

As a partial conclusion, this mag preprocessing (clahe with the 0.01 clip_limit) doesn't seem relevant as there is no improvement what so ever on any metric, but actually a decrease in some. The phase preprocessing seems more relevant but should be approach with care, further investigation is needed.

For the record, here is a visual demonstration of the applied preprocessing. 
Original images : 

<img width="619" height="352" alt="Image" src="https://github.com/user-attachments/assets/1e4ecaaf-d38c-464f-a339-2a6087e3c8cc" />

Preprocessed images : 

<img width="696" height="352" alt="Image" src="https://github.com/user-attachments/assets/e829adb9-9495-4953-87ad-f3bf41857c46" />

The preprocessing of the noisy phase image created strong artifacts in the background while the mag preprocessing introduced a slight noise.

Next step : 
Without necessarily changing the types of processing used, we could try to clean the images of the noise in the background to limit the apparition of artifacts.

**Comment by Luugaaa** [comment]
# Update on the Preprocessing
The current preprocessing steps are: 
- Otsu Thresholding to remove the background on the magnitude
- The mask obtained with the Otsu Thresholding on the magnitude is applied to the phase to remove the background
- On the magnitude, the CLAHE is applied to improve the luminance balance
- On the phase, contrast stretching is used to improve the visibility of the gray matter
 
**From the original images...:**

<img width="252" height="383" alt="Image" src="https://github.com/user-attachments/assets/8ba045e6-b5da-40fe-a7b4-26f79ac1412c" />

<img width="252" height="383" alt="Image" src="https://github.com/user-attachments/assets/926fbd78-27fa-4080-aff9-c56a453571d0" />

**This yields the following images :** 

<img width="252" height="383" alt="Image" src="https://github.com/user-attachments/assets/b8c95dd2-05df-488e-8e03-cee77fd3e90c" />

<img width="252" height="383" alt="Image" src="https://github.com/user-attachments/assets/657e75c5-4696-40e7-91bc-c664aff65fb8" />


We now have the full dataset (around 12 subjects were labeled for training and validation, out of 25). With this full dataset, we conducted ablation studies to test the effectiveness of the preprocessing. This effectiveness was hinted by a test ran before on the first _limited_ dataset (1 subject) (not uploaded on GitHub unfortunately). 

Below are the dice scores obtained on each class for the validation set (averaged on 5 folds). The curves are shown to show if there is potential improvement if the training were to go on. 

The colors are : 
- **black:**  No preprocessing
- **red:** Otsu preprocessing (background removal, always on both modalities) 
- **cyan:** Otsu + mag preprocessing (CLAHE)
- **orange:** Otsu + phase preprocessing (Contrast Stretching)

Note that the same scale is used for all classes. 

## Results 

### Class 1: White Matter

<img width="735" height="618" alt="Image" src="https://github.com/user-attachments/assets/cfa65a03-4155-43e6-b5c1-7efd177fea7a" />

**Interpretation:** The models trained with the mag preprocessing (cyan) or the phase preprocessing (orange) are _slightly_ underperforming. 

### Class 2: Gray Matter

<img width="735" height="618" alt="Image" src="https://github.com/user-attachments/assets/41e50aa4-6868-4727-82c5-f78b196f712d" />

**Interpretation:** No significant difference. 

### Class 3: Lesions on White Matter

<img width="735" height="618" alt="Image" src="https://github.com/user-attachments/assets/95cd5285-8d0c-4310-aa5d-1213a5acceac" />

**Interpretation:** The models trained with the mag preprocessing are underperforming. No other significant difference. 

### Class 4: Lesions on Gray Matter

<img width="735" height="618" alt="Image" src="https://github.com/user-attachments/assets/e3bdf72a-78e9-45df-ae80-30421c423e76" />

**Interpretation:** This is the most discriminative class, the models trained with only the Otsu thresholding (red) are performing better than the other ones, and the slope also shows that this metric could improve with a longer training. The models trained with mag prepro are, again, slightly underperforming, while the models trained with phase prepro (orange) only show a very small improvement compared to the base models (black).

## Conclusion 

The mag preprocessing is hindering the training, the phase preprocessing is only very slightly improving the results, while the Otsu thresholding seems to have a genuinely good impact on the training when it comes to segmenting lesions on gray matter (a challenging task even for the human eye). For this reason, we make the decision to discard the mag and phase preprocessing to only use the Otsu thresholding (background removal) from now on.
